### PR TITLE
Add device and resource info in external profilers

### PR DESCRIPTION
**Pull request by astachowiczhabana**
Summary: External activities could originate from external hardware.
DeviceInfo and ResourceInfo classes allows to specify hardware
information but they are not visible in public API. This change:
- makes DeviceInfo and ResourceInfo part of public API
- adds new methods in IActivityProfilerSession allowing external
  profilers communicate DeviceInfo and ResourceInfo

### CODE REVIEW

**Comment by wconstab** [comment]
@robieta same q here

File: libkineto/src/CuptiActivityProfiler.h [hunk 1]
```diff
@@ -170,7 +170,7 @@ class CuptiActivityProfiler {
     if (resourceInfo_.find({pid, tid}) == resourceInfo_.end()) {
       resourceInfo_.emplace(
           std::make_pair(pid, tid),
-          ActivityLogger::ResourceInfo(
+          libkineto::ResourceInfo(
```
Comment (briancoutinho) [inline libkineto/src/CuptiActivityProfiler.h:173]: isn't it already in libkineto namespace, may not be right on this . cc @aaronenyeshi there was some detail about KINETO_NAMESPACE that I'm missing here.

File: libkineto/src/CuptiActivityProfiler.cpp [hunk 2]
```diff
@@ -800,6 +800,18 @@ void CuptiActivityProfiler::finalizeTrace(const Config& config, ActivityLogger&
     logger.handleResourceInfo(resource, captureWindowStartTime_);
   }
 
+  for (auto &session : sessions_){
+    auto device_info = session->getDeviceInfo();
+    if (device_info != nullptr){
```
Comment (briancoutinho) [inline libkineto/src/CuptiActivityProfiler.cpp:805]: nit
`if((auto device_info = session->getDeviceInfo())) {`

File: libkineto/src/CuptiActivityProfiler.cpp [hunk 3]
```diff
@@ -800,6 +800,18 @@ void CuptiActivityProfiler::finalizeTrace(const Config& config, ActivityLogger&
...
+      logger.handleDeviceInfo(*device_info, captureWindowStartTime_);
+    }
+
+    auto resource_infos = session->getResourceInfos();
```
Comment (briancoutinho) [inline libkineto/src/CuptiActivityProfiler.cpp:809]: similar to above

File: libkineto/include/IActivityProfiler.h [hunk 4]
```diff
@@ -59,6 +85,12 @@ class IActivityProfilerSession {
   // processes trace activities using logger
   virtual void processTrace(ActivityLogger& logger) = 0;
 
+  // returns device info used in this trace, could be nullptr
+  virtual std::unique_ptr<DeviceInfo> getDeviceInfo() = 0;
```
[the remark this answers is not in the document]
Comment (astachowiczhabana) [reply libkineto/include/IActivityProfiler.h:89]: std::optinal is a C++17 feature, if I am not mistaken Kineto currently uses C++14:

libkineto/CMakeLists.txt:
...
 set_target_properties(kineto_base kineto_api PROPERTIES
      CXX_STANDARD 14
      CXX_STANDARD_REQUIRED YES
      CXX_EXTENSIONS NO)

set(KINETO_COMPILE_OPTIONS "-DKINETO_NAMESPACE=libkineto")
list(APPEND KINETO_COMPILE_OPTIONS "-DFMT_HEADER_ONLY")
if(NOT MSVC)
  list(APPEND KINETO_COMPILE_OPTIONS "-std=c++14")
else()
  list(APPEND KINETO_COMPILE_OPTIONS "/std:c++14")
  list(APPEND KINETO_COMPILE_OPTIONS "-DWIN32_LEAN_AND_MEAN")
  list(APPEND KINETO_COMPILE_OPTIONS "-DNOGDI")
endif()
...

File: libkineto/src/CuptiActivityProfiler.cpp (hunk 2 above)
Comment (astachowiczhabana) [reply libkineto/src/CuptiActivityProfiler.cpp:805]: if statement with initializer is a C++17 feature, if I am not mistaken Kineto currently uses C++14:

libkineto/CMakeLists.txt:
...
set_target_properties(kineto_base kineto_api PROPERTIES
CXX_STANDARD 14
CXX_STANDARD_REQUIRED YES
CXX_EXTENSIONS NO)

set(KINETO_COMPILE_OPTIONS "-DKINETO_NAMESPACE=libkineto")
list(APPEND KINETO_COMPILE_OPTIONS "-DFMT_HEADER_ONLY")
if(NOT MSVC)
list(APPEND KINETO_COMPILE_OPTIONS "-std=c++14")
else()
list(APPEND KINETO_COMPILE_OPTIONS "/std:c++14")
list(APPEND KINETO_COMPILE_OPTIONS "-DWIN32_LEAN_AND_MEAN")
list(APPEND KINETO_COMPILE_OPTIONS "-DNOGDI")
endif()
...

File: libkineto/src/CuptiActivityProfiler.cpp (hunk 3 above)
Comment (astachowiczhabana) [reply libkineto/src/CuptiActivityProfiler.cpp:809]: thanks for the comment, but the situation here is different: getResourceInfos returns vector

File: libkineto/src/CuptiActivityProfiler.h (hunk 1 above)
Comment (astachowiczhabana) [reply libkineto/src/CuptiActivityProfiler.h]: thx, removed it

File: libkineto/src/CuptiActivityProfiler.cpp (hunk 2 above)
Comment (briancoutinho) [reply libkineto/src/CuptiActivityProfiler.cpp:805]: ah sorry, wasn't aware of that

File: libkineto/include/IActivityProfiler.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (aaronenyeshi) [reply libkineto/include/IActivityProfiler.h:89]: Right, PyTorch is still on C++14, so we are limited to that.

File: libkineto/src/output_membuf.h [hunk 5]
```diff
@@ -97,7 +97,7 @@ class MemoryTraceLogger : public ActivityLogger {
   std::vector<const ITraceActivity*> activities_;
   std::vector<std::unique_ptr<const ITraceActivity>> wrappers_;
   std::vector<std::pair<DeviceInfo, int64_t>> deviceInfoList_;
-  std::vector<std::pair<ResourceInfo, int64_t>> resourceInfoList_;
+  std::vector<std::pair<libkineto::ResourceInfo, int64_t>> resourceInfoList_;
```
Comment (aaronenyeshi) [inline libkineto/src/output_membuf.h:100]: I don't think you need this namespace. Also how does DeviceInfo not require it but they are both defined in the same namespace?

File: libkineto/include/IActivityProfiler.h [hunk 6]
```diff
@@ -59,6 +85,12 @@ class IActivityProfilerSession {
...
+
+  // return resource info used int this trace, could be empty
+  virtual std::vector<ResourceInfo> getResourceInfos() = 0;
```
Comment (aaronenyeshi) [inline libkineto/include/IActivityProfiler.h:92]: These two functions, getDeviceInfo and getResourceInfos have been made pure virtual, but it doesn't seem to be defined in the derived classes, CuptiRangeProfilerSession, MockProfilerSession, and GlowActivityProfilerSesison. Wouldn't this break those 3 sessions?

**Review by aaronenyeshi — CHANGES_REQUESTED** [review]
Please see my reviews added. Overall, the change is fine, but we are missing override stubs in derived classes.

File: libkineto/include/IActivityProfiler.h (hunk 6 above)
Comment (astachowiczhabana) [reply libkineto/include/IActivityProfiler.h:95]: Thank you, fixed it.

File: libkineto/src/output_membuf.h (hunk 5 above)
Comment (astachowiczhabana) [reply libkineto/src/output_membuf.h]: yes, fixed

**Review by aaronenyeshi — APPROVED** [review]
Imported and fixed a few things I mentioned internally.

**Comment by aaronenyeshi** [comment]
Yes, those CI errors were fixed, and the patch has been landed.

**Pull request closed without merging**